### ISSUE TITLE
ECDHE ciphers may pose performance concerns

### ISSUE THREAD

**Issue by Miciah** [openshift/origin#4417]
The use of ECDHE ciphers for TLS connections with parallelism may pose performance problems.  https://github.com/openshift/origin/pull/2940 describes a problem we had with some test cases for the F5 router that create a test TLS server and issue some REST requests (in the order of several score) against that server.  The total runtime for the F5 router tests using `-p=4 -race` was around 33 seconds.  Without `-p=4` or without `-race`, the runtime was about 2 seconds.  We changed the test cases to configure the TLS server to prefer non-ECDHE ciphers (https://github.com/openshift/origin/pull/4414), and that brought the runtime to 2 seconds even when using `-p=4 -race`.

We need to investigate (1) the performance around Golang's ECDHE implementation generally to determine whether it may pose problems for production endpoints and (2) what the implications of using non-ECDHE ciphers are.

**Comment by liggitt** [comment]
with the ecdhe workaround removed:

go 1.4.3:
```
$ GOMAXPROCS=4 DETECT_RACES=true hack/test-go.sh pkg/router/f5 -p=4
ok  	github.com/openshift/origin/pkg/router/f5	34.488s	coverage: 72.9% of statements
hack/test-go.sh succeeded after 49 seconds
```

go 1.5.3:
```
$ GOMAXPROCS=4 DETECT_RACES=true hack/test-go.sh pkg/router/f5 -p=4
ok  	github.com/openshift/origin/pkg/router/f5	17.871s	coverage: 72.9% of statements
hack/test-go.sh succeeded after 46 seconds
```

go 1.6:
```
$ GOMAXPROCS=4 DETECT_RACES=true hack/test-go.sh pkg/router/f5 -p=4
ok  	github.com/openshift/origin/pkg/router/f5	3.974s	coverage: 72.9% of statements
hack/test-go.sh succeeded after 31 seconds
```

**Comment by liggitt** [comment]
fixed in 1.6
